GoogleForCreators/web-stories-wp
# Smart text style: ignore selected element in contrast calculation

## Context

When generating an image for contrast calculation, the text element itself is also included into the image, that causes unexpected and inaccurate results.

## Summary

Improve how the pixel data for contrast calculation is generated before calculating the contrast by preventing the text element from being included in the pixel data that we use to calculate the contrast. 

## Relevant Technical Choices

The [html-to-image](https://github.com/bubkoo/html-to-image) library we use to extract the pixels for contrast calculation provides a filter option which we can use to filter out nodes from the DOM which we want to exclude from the generated canvas.



## To-do

n/a


## User-facing changes
Before
![Screen-Recording-2021-11-16-at-0](https://user-images.githubusercontent.com/276316/141925088-d4bc1399-a1b8-4b4b-8d51-56fd1d1f4363.gif)


After
![Screen-Recording-2021-11-16-at-0 (1)](https://user-images.githubusercontent.com/276316/141925047-d12e2c4d-e70b-4c88-8b99-c96376b4548b.gif)

## Testing Instructions






This PR can be tested by following these steps:

1. Add a text element on a white background
2. Click on the "Magic wand" button for auto-styling.
3. The text should stay black as it is without any highlight.
4. Verifying that auto-styling still works correctly in other cases e.g. Adds a highlight incase of an applicable background image


## Reviews

### Does this PR have a security-related impact?

No


### Does this PR change what data or activity we track or use?

No


### Does this PR have a legal-related impact?

No


## Checklist



- [] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))

---



Fixes #9546
